### PR TITLE
[PrintAsObjC] Remove empty extensions from Swift compatibility headers.

### PR DESCRIPTION
**Pull request by calebd**
This change removes empty extension interfaces from generated Swift compatibility headers where empty extensions add neither public members or protocol conformances.

Resolves [SR-2772](https://bugs.swift.org/browse/SR-2772).

### CODE REVIEW

**Comment by jrose-apple** [comment]
This is definitely an improvement over the status quo, but it's not quite the whole SR-2772. Note that in that example there *is* a member in the extension; it's just not one that's printed. There could be a variety of reasons for this, the top two being access control and whether it's actually ObjC-compatible.

File: test/PrintAsObjC/extensions.swift [hunk 1]
```diff
@@ -95,8 +118,7 @@ extension GenericClass {
 class NotObjC {}
 extension NotObjC {}
 
-// CHECK-LABEL: @interface NSObject (SWIFT_EXTENSION(extensions))
-// CHECK-NEXT: @end
+// NEGATIVE-NOT: @interface NSObject (SWIFT_EXTENSION(extensions))
 // NEGATIVE-NOT: @interface NSObject{{$}}
 // NEGATIVE-NOT: @class NSObject
 extension NSObject {}
```
Comment (jrose-apple) [inline test/PrintAsObjC/extensions.swift]: This part is supposed to be a minimal test that we don't try to *define* an imported class. Can you put a member in instead, like you did with A2?

File: lib/PrintAsObjC/PrintAsObjC.cpp [hunk 2]
```diff
@@ -329,7 +329,33 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     os << "@end\n";
   }
 
+  bool isEmptyExtensionDecl(ExtensionDecl *ED) {
+    auto members = ED->getMembers();
+    SmallVector<Decl *, 4> includedMembers;
+    std::copy_if(members.begin(), members.end(),
```
Comment (jrose-apple) [inline lib/PrintAsObjC/PrintAsObjC.cpp]: You can do even better by using [`std::any_of`](http://en.cppreference.com/w/cpp/algorithm/all_any_none_of).

**Review by jrose-apple — COMMENTED** [review]
Thanks! This is more what we want. :-) A few comments, still.

File: test/PrintAsObjC/extensions.swift (hunk 1 above)
Comment (calebd) [inline test/PrintAsObjC/extensions.swift]: Makes sense. In `master` it looks like the test first asserts that there is an extension, then asserts that there is no class defined. Due to the sorted nature of this header would it be more correct to first assert that the class declaration is not there followed by the extension check?

Comment (jrose-apple) [inline test/PrintAsObjC/extensions.swift]: The CHECK and NEGATIVE lines are being checked in two different passes, so it's not going to be a problem in practice. I'm happy to have them reordered, though; that would match the cases where we *do* expect the extended class to be defined first.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Merged. Thanks, Caleb!

**Pull request merged**